### PR TITLE
add checks to test dirs (closes #347)

### LINKED ISSUE #347: unhelpful test suite error when CH_TEST_*DIR does not exist
If you have `CH_TEST_*DIR` arguments set but the directories do not exist the test suite will give this unhelpful error:

```
$ make test-build
bats build.bats build_auto.bats build_post.bats
/home/jogas/git/charliecloud/test/bats/libexec/bats-exec-suite: line 20: let: count+=: syntax error: operand expected (error token is "+=")
make: *** [test-build] Error 1
```

If the directories do not exist the error message should reflect that. Alternatively, the `CH_TEST_*DIR` directories should just be created if they do not exist and error if they cannot be created at the specified path. FWIW, I much prefer the latter.

### PR DESCRIPTION
**Pull request by j-ogas**
Addresses #347.

### CODE REVIEW

File: test/common.bash [hunk 1]
```diff
@@ -148,8 +153,8 @@ ch_version_docker=$(echo "$ch_version" | tr '~+' '--')
 # despite the admonition in the man page, because it's more portable [1].
 #
 # [1]: https://unix.stackexchange.com/a/136527
-ch_imgdir=$(readlink -ef "$CH_TEST_IMGDIR")
-ch_tardir=$(readlink -ef "$CH_TEST_TARDIR")
+ch_imgdir=$(readlink -ef "$CH_TEST_IMGDIR") || fatal "CH_TEST_IMGDIR does not exist"
+ch_tardir=$(readlink -ef "$CH_TEST_TARDIR") || fatal "CH_TEST_TARDIR does not exist"
```
Comment (reidpr) [inline test/common.bash]: I think the thing to do here is make a wrapper function for `readlink(1)` that complains if the directory doesn't exist.

It should also report which directory does not exist, not just the variable containing the name of the nonexistent directory.

File: test/common.bash [hunk 2]
```diff
@@ -30,6 +30,11 @@ env_require () {
     fi
 }
 
+fatal () {
+    printf '%s\n\n' "$1" >&2
+    exit 1
+}
+
```
Comment (reidpr) [inline test/common.bash]: I like abstracting this into a function.

I'd like for this PR to change the other error exits to use it too.

**Comment by j-ogas** [comment]
I am having issues (stumped) with including `CH_TEST_PERMDIRS`. Bats takes it upon itself to exit despite the code executing properly. If you follow the output of the execution below, it seems that `common.bash` executes entirely and then once bats enters its `bats_test_function`, it throws a error.

For example:
```
$ export CH_TEST_TARDIR=/tmp
$ export CH_TEST_IMGDIR=/var/tmp
$ export CH_TEST_PERMDIRS='/tmp /var/tmp'
$ make test
bats build.bats build_auto.bats build_post.bats
+++ (( i=1 ))
+++ (( i<=11 ))
++++ cut -d ' ' -f1
+++ d=/tmp
+++ [[ -n /tmp ]]
+++ read_link -e /tmp CH_TEST_PERMDIRS
+++ readlink -e /tmp
+++ echo /tmp
+++ (( i++ ))
+++ (( i<=11 ))
++++ cut -d ' ' -f2
+++ d=/var/tmp
+++ [[ -n /var/tmp ]]
+++ read_link -e /var/tmp CH_TEST_PERMDIRS
+++ readlink -e /var/tmp
+++ echo /var/tmp
+++ (( i++ ))
+++ (( i<=11 ))
++++ cut -d ' ' -f3
+++ d=/tmp/delete
+++ [[ -n /tmp/delete ]]
+++ read_link -e /tmp/delete CH_TEST_PERMDIRS
+++ readlink -e /tmp/delete
+++ echo /tmp/delete
+++ (( i++ ))
+++ (( i<=11 ))
++++ cut -d ' ' -f4
+++ d=
+++ [[ -n '' ]]
+++ break
+++ bash -c 'set -e; [[ 1 = 0 ]]; exit 0'
+++++ dirname /home/jogas/upstream/test/common.bash
++++ cd /home/jogas/upstream/test/bin
++++ pwd
+++ ch_bin=/home/jogas/upstream/test/bin
++++ read_link -f /home/jogas/upstream/test/bin ch_bin
++++ readlink -f /home/jogas/upstream/test/bin
++++ echo /home/jogas/upstream/test/bin
+++ ch_bin=/home/jogas/upstream/test/bin
+++ export PATH=/home/jogas/upstream/test/bin:/usr/libexec:/home/jogas/git/charliecloud/test/bats/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
+++ PATH=/home/jogas/upstream/test/bin:/usr/libexec:/home/jogas/git/charliecloud/test/bats/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
++++ command -v ch-run
+++ ch_runfile=/home/jogas/upstream/test/bin/ch-run
++++ ch-build --libexec-path
+++ ch_libexec=/home/jogas/upstream/test/bin
+++ [[ -x /home/jogas/upstream/test/bin/ch-run ]]
++++ ch-run --version
+++ ch_version=0.9.7~pre+epel-package.d9ad4b6
++++ echo 0.9.7~pre+epel-package.d9ad4b6
++++ tr '~+' --
+++ ch_version_docker=0.9.7-pre-epel-package.d9ad4b6
++++ read_link -e /tmp/delete CH_TEST_IMGDIR
++++ readlink -e /tmp/delete
++++ echo /tmp/delete
+++ ch_imgdir=/tmp/delete
++++ read_link -e /tmp/delete CH_TEST_TARDIR
++++ readlink -e /tmp/delete
++++ echo /tmp/delete
+++ ch_tardir=/tmp/delete
+++ mount
+++ grep -Fq /tmp/delete
+++ ch_tag=NO_TAG_SET
+++ ch_img=/tmp/delete/NO_TAG_SET
+++ ch_tar=/tmp/delete/NO_TAG_SET.tar.gz
+++ ch_ttar=/tmp/delete/chtest.tar.gz
+++ ch_timg=/tmp/delete/chtest
+++ btnew=/tmp/bats.tmp.jogas
+++ mkdir -p /tmp/bats.tmp.jogas
+++ chmod 700 /tmp/bats.tmp.jogas
+++ export BATS_TMPDIR=/tmp/bats.tmp.jogas
+++ BATS_TMPDIR=/tmp/bats.tmp.jogas
++++ stat -c %a /tmp/bats.tmp.jogas
+++ [[ 700 = \7\0\0 ]]
+++ [[ NO_TAG_SET = *\-\m\p\i\c\h ]]
+++ ch_mpi=openmpi
+++ [[ -f /etc/opt/cray/release/cle-release ]]
+++ ch_cray=
+++ [[ -n '' ]]
+++ ch_nodes=1
++++ getconf _NPROCESSORS_ONLN
+++ ch_cores_node=12
+++ ch_cores_total=12
+++ [[ openmpi = mpich ]]
+++ ch_mpirun_np=--use-hwthread-cpus
+++ ch_unslurm=
+++ [[ -n '' ]]
+++ ch_multinode=
+++ command -v mpirun
+++ ch_multiprocess=
+++ ch_mpirun_node=
+++ ch_mpirun_core=false
+++ ch_mpirun_2=false
+++ ch_mpirun_2_1node=false
+++ command -v docker
+++ [[ -z full ]]
+++ [[ full != quick ]]
+++ [[ full != standard ]]
+++ [[ full != full ]]
+++ command -v sudo
+++ sudo -v
+++ ch_have_sudo=yes
++ bats_test_function test_create_tarball_directory_if_needed
++ local test_name=test_create_tarball_directory_if_needed
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_create_tarball_directory_if_needed
++ bats_test_function test_documentations_build
++ local test_name=test_documentations_build
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_documentations_build
++ bats_test_function test_version_number_seems_sane
++ local test_name=test_version_number_seems_sane
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_version_number_seems_sane
++ bats_test_function test_executables_seem_sane
++ local test_name=test_executables_seem_sane
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_executables_seem_sane
++ bats_test_function test_lint_shell_scripts
++ local test_name=test_lint_shell_scripts
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_lint_shell_scripts
++ bats_test_function test_proxy_variables
++ local test_name=test_proxy_variables
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_proxy_variables
++ bats_test_function test_ch-2dbuild2dir
++ local test_name=test_ch-2dbuild2dir
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_ch-2dbuild2dir
++ bats_test_function test_ch-2dpull2tar
++ local test_name=test_ch-2dpull2tar
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_ch-2dpull2tar
++ bats_test_function test_ch-2dpull2dir
++ local test_name=test_ch-2dpull2dir
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_ch-2dpull2dir
++ bats_test_function test_sotest_executable_works
++ local test_name=test_sotest_executable_works
++ BATS_TEST_NAMES["${#BATS_TEST_NAMES[@]}"]=test_sotest_executable_works
+ '[' -n 1 ']'
+ echo 10
+ bats_cleanup_preprocessed_source
+ rm -f /tmp/bats.120043.src
/usr/libexec/bats-exec-suite: line 20: let: count+=/tmp
/var/tmp
/tmp/delete
10: syntax error: operand expected (error token is "/tmp
/var/tmp
/tmp/delete
10")

**Comment by j-ogas** [comment]
Can't reproduce original bug. Closing.

**Pull request closed without merging**